File: mastra_agent/src/bffToolAdapter.ts
import { createTool } from '@mastra/core/tools';
import { z } from 'zod';

export interface ToolSchema {
  name: string;
  description: string;
  inputSchema: Record<string, unknown>;
}

export interface RunCtx {
  bffToolUrl: string;
  bffInternalSecret: string;
  sessionId: string;
}

export class BffToolError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'BffToolError';
  }
}

interface JsonSchemaProperty {
  type?: string;
}

interface JsonSchemaObject {
  properties?: Record<string, JsonSchemaProperty>;
}

export function buildBffTools(schemas: ToolSchema[], runCtx: RunCtx) {
  return schemas.map((schema) => _makeTool(schema, runCtx));
}

function _makeTool(schema: ToolSchema, runCtx: RunCtx) {
  const props = (schema.inputSchema as JsonSchemaObject).properties ?? {};
  const zodShape: Record<string, z.ZodTypeAny> = {};
  for (const [key, val] of Object.entries(props)) {
    zodShape[key] = val.type === 'number' ? z.number().optional() : z.string().optional();
  }

  const executeImpl = async (args: Record<string, unknown>) => {
    const resp = await fetch(runCtx.bffToolUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'x-internal-gateway-secret': runCtx.bffInternalSecret,
        'x-session-id': runCtx.sessionId,
      },
      body: JSON.stringify({ tool: schema.name, args, sessionId: runCtx.sessionId }),
    });
    if (!resp.ok) {
      const text = await resp.text();
      throw new BffToolError(`BFF returned HTTP ${resp.status}: ${text.slice(0, 200)}`);
    }
    const data = (await resp.json()) as Record<string, unknown>;
    return (data.result as Record<string, unknown> | undefined) ?? data;
  };

  return createTool({
    id: schema.name,
    description: schema.description,
    inputSchema: z.object(zodShape),
    execute: async (inputData: Record<string, unknown>) => executeImpl(inputData),
  });
}

File: pydantic_agent/src/bff_tool_adapter.py
from __future__ import annotations
from typing import Any
import httpx
from pydantic_ai import RunContext
from pydantic_ai.tools import Tool
from .models import BffDeps


class BffToolError(Exception):
    pass


def build_tool_functions(tool_schemas: list[dict]) -> list[Tool]:
    tools = []
    for schema in tool_schemas:
        tools.append(_make_tool(schema))
    return tools


def _make_tool(schema: dict) -> Tool:
    name: str = schema["name"]
    description: str = schema["description"]
    properties: dict = schema.get("inputSchema", {}).get("properties", {})
    param_names: list[str] = list(properties.keys())

    async def tool_fn(ctx: RunContext[BffDeps], **kwargs: Any) -> Any:
        args = {k: kwargs[k] for k in param_names if k in kwargs}
        async with httpx.AsyncClient() as client:
            resp = await client.post(
                ctx.deps.bff_tool_url,
                json={"tool": name, "args": args, "sessionId": ctx.deps.session_id},
                headers={
                    "x-internal-gateway-secret": ctx.deps.bff_internal_secret,
                    "x-session-id": ctx.deps.session_id,
                },
                timeout=30.0,
            )
        if resp.status_code != 200:
            raise BffToolError(f"BFF returned HTTP {resp.status_code}: {resp.text[:200]}")
        data = resp.json()
        return data.get("result", data)

    tool_fn.__name__ = name
    tool_fn.__doc__ = description
    return Tool(tool_fn, name=name, description=description, takes_ctx=True)
